Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py

```diff
@@ -0,0 +1,345 @@
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+import json
+import urllib3
+import traceback
+from typing import Any, Dict, List, Union
+from MicrosoftApiModule import *
+
+# Disable insecure warnings
+urllib3.disable_warnings()
+
+''' CONSTANTS '''
+
+UPSERT_PARAMS = {'resource_id': 'id', 'policy_settings': 'properties.policySettings',
+                 'location': 'location', 'custom_rules': 'properties.customRules',
+                 'tags': 'tags', 'managed_rules': 'properties.managedRules'}
+
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+API_VERSION = '2020-05-01'
+
+''' CLIENT CLASS '''
+
+
+class AzureWAFClient:
+    @logger
+    def __init__(self, app_id, subscription_id, resource_group_name, verify, proxy):
+
+        if '@' in app_id:
+            app_id, refresh_token = app_id.split('@')
+            integration_context = demisto.getIntegrationContext()
+            integration_context.update(current_refresh_token=refresh_token)
+            demisto.setIntegrationContext(integration_context)
+        base_url = f'https://management.azure.com/subscriptions/{subscription_id}/'
+        self.subscription_id = subscription_id
+        self.resource_group_name = resource_group_name
+        client_args = client_args = {
+            'self_deployed': True,  # We always set the self_deployed key as True because when not using a self
+                                    # deployed machine, the DEVICE_CODE flow should behave somewhat like a self deployed
+                                    # flow and most of the same arguments should be set, as we're !not! using OProxy.
+            'auth_id': app_id,
+            'token_retrieval_url': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/token',
+            'grant_type': DEVICE_CODE,  # disable-secrets-detection
+            'base_url': base_url,
+            'verify': verify,
+            'proxy': proxy,
+            'resource': 'https://management.core.windows.net',   # disable-secrets-detection
+            'scope': 'https://management.azure.com/user_impersonation offline_access user.read',
+            'ok_codes': (200, 201, 202, 204)
+        }
+
+        self.ms_client = MicrosoftClient(**client_args)
+
+    @logger
+    def http_request(self, method: str, url_suffix: str = None, full_url: str = None, params: Dict = None,
+                     data: Dict = None, resp_type: str = 'json', return_empty_response: bool = False):
+        if not params:
+            params = {}
```

Comment:
maybe we can se it in the func args?

Reply:
It is bad practice as far as I know to put mutable object in a function's arguments, though in this specific case it does not matter, I prefer to keep it as None.